riscv/src/elf.rs

```diff
@@ -0,0 +1,1229 @@
+
+    assert_eq!(elf.dynrels.len(), 0, "Unsupported relocation type!");
+}
+
+fn static_relocate_data_sections(
+    elf: &Elf,
+    address_map: &AddressMap,
+    data_map: &mut BTreeMap<u32, Data>,
+    referenced_text_addrs: &mut BTreeSet<u32>,
+) {
+    // In non-PIE files, we need to use the linking relocation table.
+    for r in elf.shdr_relocs.iter().flat_map(|(_, relocs)| relocs.iter()) {
+        let addr = r.r_offset as u32;
+        if !address_map.is_in_data_section(addr) {
+            // Relocation of the text section has already been handled in instruction lifting.
+            continue;
+        }
+
+        // We only support the R_RISCV_32 relocation type for the data section:
+        assert_eq!(r.r_type, R_RISCV_32, "Unsupported relocation type!");
+
+        let Entry::Occupied(mut entry) = data_map.entry(r.r_offset as u32) else {
+            panic!("Unexpected 0 in relocated data entry!");
+        };
+
+        let Data::Value(original_addr) = *entry.get() else {
+            panic!("Related entry already replaced with a label!");
+        };
+
+        if address_map.is_in_text_section(original_addr) {
+            entry.insert(Data::TextLabel(original_addr));
+
+            // We also need to add the referenced address to the list of text
+            // addresses, so we can generate the label.
+            referenced_text_addrs.insert(original_addr);
+        }
+    }
+}
+
+/// Index the symbols by their addresses.
+struct SymbolTable(HashMap<u32, String>);
+
+impl SymbolTable {
+    fn new(elf: &Elf) -> SymbolTable {
+        // TODO: read the symbols from the debug information to be more comprehensive.
+        let mut deduplicator = HashMap::new();
+        for sym in elf.syms.iter() {
+            if sym.st_name == 0 || (sym.st_type() != STT_OBJECT && sym.st_type() != STT_FUNC) {
+                continue;
+            }
+            deduplicator.insert(elf.strtab[sym.st_name].to_string(), sym.st_value as u32);
+        }
+
+        Self(
+            deduplicator
+                .into_iter()
+                .map(|(name, addr)| (addr, name))
+                .collect(),
+        )
+    }
+
+    /// Get the symbol if the address had one.
+    fn try_get(&self, addr: u32) -> Option<&str> {
+        self.0.get(&addr).map(|name| name.as_str())
+    }
+
+    /// Get the symbol or a default label formed from the address value.
+    fn get(&self, addr: u32) -> Cow<str> {
+        self.0
+            .get(&addr)
+            .map(|name| Cow::Borrowed(name.as_str()))
+            .unwrap_or_else(|| Cow::Owned(format!("L{addr:08x}")))
+    }
+
+    /// Get the symbol or a default label formed from the address value.
+    fn get_as_string(&self, addr: u32) -> String {
+        self.get(addr).into_owned()
+    }
+}
+
+impl RiscVProgram for ElfProgram {
+    fn take_source_files_info(&mut self) -> impl Iterator<Item = crate::code_gen::SourceFileInfo> {
+        // TODO: read the source files from the debug information.
```

Comment:
here or later?

Reply:
I removed this TODO because we already have one on the topic of debug symbols